components/canvas/webgl_mode/inprocess.rs

```diff
@@ -94,13 +107,34 @@ impl WebGLExternalImageApi for WebGLExternalImages {
         self.webgl_channel
             .send(WebGLMsg::Lock(ctx_id, self.lock_channel.0.clone()))
             .unwrap();
-        let (image_id, size, gl_sync) = self.lock_channel.1.recv().unwrap();
-        // The next glWaitSync call is run on the WR thread and it's used to synchronize the two
-        // flows of OpenGL commands in order to avoid WR using a semi-ready WebGL texture.
-        // glWaitSync doesn't block WR thread, it affects only internal OpenGL subsystem.
-        self.webrender_gl
-            .wait_sync(gl_sync as gl::GLsync, 0, gl::TIMEOUT_IGNORED);
-        (image_id, size)
+        let (image_id, size, io_surface_id, gl_sync) = self.lock_channel.1.recv().unwrap();
+
+        let texture_id = match io_surface_id {
+            Some(io_surface_id) => {
+                let texture_id = match self.textures.entry(io_surface_id) {
+                    Entry::Occupied(o) => *o.get(),
+                    Entry::Vacant(v) => {
+                        let texture_id = self.webrender_gl.gen_textures(1)[0];
+                        self.webrender_gl
+                            .bind_texture(gl::TEXTURE_RECTANGLE, texture_id);
+                        let io_surface = io_surface::lookup(io_surface_id);
+                        io_surface.bind_to_gl_texture(size.width, size.height);
+                        *v.insert(texture_id)
+                    },
+                };
+                texture_id
```

Comment:
There's no locking or sync'ing needed here? How does this avoid race conditions between the WebGL thread and the WebRender thread?

Reply:
So far the triple buffering result worked for me, because the time interval between `Lock` and `Unlock` messages was lesser than switching texture targets on the WebGL thread two times. But I can agree adding it would be make the code more safer, I will look into that.